File: Data Structures & Algorithms/reverse-a-linked-list/submission-6.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None
        newHead = head
        nextElem = head.next
        head.next = None
        if nextElem:
            newHead = self.reverseList(nextElem)
            nextElem.next = head
        return newHead

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-6.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:

    ListNode* mergetwo(ListNode* l1, ListNode* l2) {
        ListNode dummy = ListNode(), * res = &dummy;
        ListNode** ptr;
        while (l1 && l2) {
            if (l1->val <= l2->val) ptr = &l1;
            else ptr = &l2;
            
            res->next = *ptr;
            *ptr = (*ptr)->next;
            res = res->next;
        }

        if (l1) res->next = l1;
        if (l2) res->next = l2;

        return dummy.next;
    }

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.size() == 0) return nullptr;
        
        ListNode* res = lists[0];
        for (int i = 1; i < lists.size(); i++) {
            res = mergetwo(res, lists[i]);
        }

        return res;
    }
};
